### ISSUE TITLE
TypeScript 7.0.2 Support

### ISSUE THREAD

**Issue by IshtarStar** [typescript-eslint/typescript-eslint#12518]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Relevant Package

typescript-eslint / @typescript-eslint/typescript-estree

### Description

I tried upgrading a Vite React TypeScript project from TypeScript 6.0.3 to TypeScript 7.0.2.

With `typescript-eslint@8.63.0`, installation via `npm ci` fails because the published peer dependency range only allows TypeScript `<6.1.0`.

If dependencies are installed with `npm install`, ESLint then crashes inside `@typescript-eslint/typescript-estree`.

I understand TypeScript 7 may not be officially supported yet. I am filing this as a TypeScript 7 support report / compatibility tracking data point.

### Versions

- Node: 24.13.0
- npm: 11.6.2
- OS: macOS Darwin arm64
- eslint: 10.6.0
- typescript-eslint: 8.63.0
- typescript: 7.0.2
- vite: 8.1.3

### package.json relevant devDependencies

```json
{
  "devDependencies": {
    "eslint": "^10.6.0",
    "typescript": "^7.0.2",
    "typescript-eslint": "^8.63.0",
    "vite": "^8.1.3"
  }
}
```
### Reproduction Steps
1. Set TypeScript to ^7.0.2
2. Run:
```
npm install
npm ci
npm run lint
```
## Actual Result
npm ci fails with:
```
npm error ERESOLVE could not resolve

While resolving: typescript-eslint@8.63.0
Found: typescript@7.0.2

Could not resolve dependency:
peer typescript@">=4.8.4 <6.1.0" from typescript-eslint@8.63.0
```
If installed via npm install, linting fails with:

```
ESLint: 10.6.0

TypeError: Cannot read properties of undefined (reading 'Cjs')
    at .../node_modules/typescript-eslint/node_modules/@typescript-eslint/typescript-estree/dist/create-program/shared.js
```
### Expected Result
Either:
* TypeScript 7 is supported by typescript-eslint, or
* TypeScript 7 unsupported usage fails with a clear compatibility warning/error instead of an internal crash.
### Additional Context
npm run build succeeds with TypeScript 7.0.2 in this project, but npm ci and npm run lint fail.

### Reproduction Repository Link

https://github.com/IshtarStar/typescript-eslint-typescript-7-repro

### Repro Steps

1. Clone the reproduction repository:

```sh
   git clone https://github.com/IshtarStar/typescript-eslint-typescript-7-repro.git
   cd typescript-eslint-typescript-7-repro
```
2. Reproduce the install failure:
```
npm ci
```
This fails because typescript-eslint@8.63.0 has a peer dependency range of typescript >=4.8.4 <6.1.0, while the project uses typescript@7.0.2.
3. Reproduce the ESLint crash after forcing installation:
```npm install --force
npm run versions
npm run lint
```
4. Observe the ESLint crash:
```
TypeError: Cannot read properties of undefined (reading 'Cjs')
    at .../@typescript-eslint/typescript-estree/dist/create-program/shared.js:59:18
```
The repository also includes a GitHub Actions workflow with two jobs that reproduce both cases automatically:
* npm ci peer dependency conflict
* ESLint crash after npm install

### Versions

| package                                | version   |
| -------------------------------------- | --------- |
| `@typescript-eslint/eslint-plugin`     | `8.63.0`  |
| `@typescript-eslint/parser`            | `8.63.0`  |
| `@typescript-eslint/scope-manager`     | `8.63.0`  |
| `@typescript-eslint/typescript-estree` | `8.63.0`  |
| `@typescript-eslint/type-utils`        | `8.63.0`  |
| `@typescript-eslint/utils`             | `8.63.0`  |
| `TypeScript`                           | `7.0.2`   |
| `ESLint`                               | `10.6.0`  |
| `node`                                 | `24.18.0` |
| `npm` | `11.16.0` |

**Comment by IshtarStar** [comment]
I added a separate branch that demonstrates a temporary workaround:

https://github.com/IshtarStar/typescript-eslint-typescript-7-repro/tree/fix/typescript-7-eslint-patch

This is not a real fix for TypeScript 7 support in typescript-eslint.

What the workaround does:

- keeps `typescript` on `^6.0.3`, so `typescript-eslint@8.63.0` still gets a supported peer dependency
- installs TypeScript 7 as an npm alias: `typescript-7: npm:typescript@^7.0.2`
- runs ESLint against the supported TypeScript 6 peer
- runs the TypeScript 7 compiler separately via `npm run tsc7`

The branch validates successfully with:
```sh
npm ci
npm run versions
npm run lint
npm run tsc7
```
This avoids the ESLint crash, but it also shows that current `typescript-eslint@8.63.0` cannot directly use `typescript@7.0.2`.
While investigating, I noticed that with `typescript@7.0.2`, `require("typescript")` only exposes:
```
version
versionMajorMinor
```
So the crash at:
```
@typescript-eslint/typescript-estree/dist/create-program/shared.js:59:18
```
happens because `typescript-estree` expects legacy TypeScript APIs such as `ts.Extension.Cjs`, which are not exposed from the default `typescript` entrypoint in TypeScript 7.

**Comment by kirkwaiblinger** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/10940

**Comment by kirkwaiblinger** [comment]
@IshtarStar typescript-eslint isn't compatible with TS 7 at this time, because there is no TS 7 API at this time. There is nothing we can do about this until TS 7 provides an API. You'll just want to set up typescript-eslint to use TS 6, for example as described in https://devblogs.microsoft.com/typescript/announcing-typescript-7-0/#running-side-by-side-with-typescript-6.0

**Comment by IshtarStar** [comment]
Thanks, understood. I found #10940 as the relevant tracking issue.

I’ll use the TS 6 + TS 7 side-by-side setup for now:
https://github.com/IshtarStar/typescript-eslint-typescript-7-repro/tree/fix/typescript-7-eslint-patch

This can be closed as currently unsupported / blocked on a TS 7 API.

**Comment by heyakyra** [comment]
Shouldn't this be planned since the API was always intended, but especially now that typescript 7.1 pre-release versions are available this ticket can be reopened and a similar pre-release version can be worked on? https://www.npmjs.com/package/typescript/v/next

Is there a way to mark it as a duplicate instead of not planned, since #10940 was filed first?
